### PR TITLE
RA-OSS-18 - Ability to style emptyText on fields

### PR DESCRIPTION
**Pull request by gitstart**
[ISSUE](https://github.com/marmelab/react-admin/issues/8951)
## Loom Video:
https://www.loom.com/share/a15ca7e68a04480e9c55fc4ab96d8308?sid=ba2c408c-4d7b-41c9-a2aa-0f2bdecdd546

### CODE REVIEW

**Comment by fzaninotto** [comment]
As this is a new feature, please rebase on `next` rather than `master`.

**Comment by fzaninotto** [comment]
You must also update the documentation in `Fields.md` and add unit tests.

**Comment by gitstart** [comment]
> You must also update the documentation in `Fields.md` and add unit tests.

PR has been rebased to `next` @fzaninotto

File: packages/ra-ui-materialui/src/field/ImageField.spec.tsx [hunk 1]
```diff
@@ -184,4 +184,20 @@ describe('<ImageField />', () => {
 
         expect(getByText('Not found')).not.toBeNull();
     });
+
+    it.each([null, undefined])(
+        'hould render the emptyText ReactNode when value is %s',
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ImageField.spec.tsx:189]:
```suggestion
        'should render the emptyText ReactNode when value is %s',
```

File: docs/Fields.md [hunk 2]
```diff
@@ -129,7 +129,7 @@ All Field components accept the following props:
...
 | [`className`](#classname)     | Optional | `string`                       | -        | A class name (usually generated by JSS) to customize the look and feel of the field element itself                                                      |
 | [`textAlign`](#textalign)     | Optional | `string`                       | 'left'   | Defines the text alignment inside a cell. Set to `right` for right alignment (e.g. for numbers)                                                         |
-| [`emptyText`](#emptytext)     | Optional | `string`                       | ''       | Defines a text to be shown when a field has no value (not supported in array fields)                                                                    |
+| [`emptyText`](#emptytext)     | Optional | `string` &#124; `ReactNode`    | ''       | Defines a text or element to be shown when a field has no value (not supported in array fields)                                                                    |
```
Comment (fzaninotto) [inline docs/Fields.md:132]:
```suggestion
| [`emptyText`](#emptytext)     | Optional | `string | ReactNode`    | ''       | Defines a text or element to be shown when a field has no value (not supported in array fields)                                                                    |
```

File: docs/Fields.md [hunk 3]
```diff
@@ -151,7 +151,10 @@ const PostList = () => (
     <List>
         <Datagrid>
             <TextField source="title" />
+            {/* emptyText can be a string */}
             <TextField source="author" emptyText="missing data" />
+            {/* you can use React Node as emptyText */}
+            <TextField source="email" emptyText={<b>missing data</>} />
```
Comment (fzaninotto) [inline docs/Fields.md:157]:
```suggestion
            <TextField source="email" emptyText={<em>missing data</em>} />
```

File: docs/ReferenceField.md [hunk 4]
```diff
@@ -76,7 +76,7 @@ With this configuration, `<ReferenceField>` wraps the user's name in a link to t
 | `source`    | Required | `string`            | -        | Name of the property to display |
 | `reference` | Required | `string`            | -        | The name of the resource for the referenced records, e.g. 'posts' |
 | `children`  | Optional | `ReactNode`         | -        | One or more Field elements used to render the referenced record |
-| `emptyText` | Optional | `string`            | ''       | Defines a text to be shown when the field has no value or when the reference is missing |
+| `emptyText` | Optional | `string` &#124; `ReactNode`    | ''       | Defines a text or element to be shown when the field has no value or when the reference is missing |
```
Comment (fzaninotto) [inline docs/ReferenceField.md:79]:
```suggestion
| `emptyText` | Optional | `string | ReactNode`    | ''       | Defines a text or element to be shown when the field has no value or when the reference is missing |
```

File: packages/ra-ui-materialui/src/field/BooleanField.spec.tsx [hunk 5]
```diff
@@ -150,6 +150,22 @@ describe('<BooleanField />', () => {
...
+            const emptyTextNode = <b>Custom Empty Text</b>;
+            render(
+                <BooleanField
+                    {...defaultProps}
+                    record={{ id: 123, published }}
+                    emptyText={emptyTextNode}
+                />
+            );
+            const renderedEmptyText = screen.getByText('Custom Empty Text');
+            expect(renderedEmptyText).toBeTruthy();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/BooleanField.spec.tsx:165]: `getByXXX` throws if the searched element is not found. This `expect` doesn't really make sense. So either use `queryByYYY` or remove the `expect`. 

See https://testing-library.com/docs/react-testing-library/cheatsheet/#queries

Same for other tests.

File: packages/ra-ui-materialui/src/field/DateField.spec.tsx [hunk 6]
```diff
@@ -194,4 +194,20 @@ describe('<DateField />', () => {
...
     });
+
+    it.each([null, undefined])(
+        'should display emptyText ReactNode prop if provided for %s value',
+        foo => {
+            const emptyTextNode = <b>Custom Empty Text</b>;
+            const { getByText } = render(
+                <DateField
+                    record={{ id: 123, foo }}
+                    source="foo.bar"
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/DateField.spec.tsx:205]:
```suggestion
                    source="foo"
```

File: packages/ra-ui-materialui/src/field/NumberField.spec.tsx [hunk 7]
```diff
@@ -126,4 +126,20 @@ describe('<NumberField />', () => {
...
     });
+
+    it.each([null, undefined])(
+        'should display emptyText ReactNode prop if provided for %s value',
+        foo => {
+            const emptyTextNode = <b>Custom Empty Text</b>;
+            const { getByText } = render(
+                <NumberField
+                    record={{ id: 123, foo }}
+                    source="foo.bar"
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/NumberField.spec.tsx:137]:
```suggestion
                    source="foo"
```

File: packages/ra-ui-materialui/src/field/UrlField.spec.tsx [hunk 8]
```diff
@@ -77,4 +77,20 @@ describe('<UrlField />', () => {
...
     });
+
+    it.each([null, undefined])(
+        'should render the emptyText ReactNode when value is %s',
+        url => {
+            const emptyTextNode = <b>Custom Empty Text</b>;
+            const { getByText } = render(
+                <UrlField
+                    record={{ id: 123, url }}
+                    className="foo"
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/UrlField.spec.tsx:88]: Remove this one, as it's useless in that case

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Note that since the value of `emptyText` is no longer a `string`, its title becomes misleading. But we won't add another prop just for that, or introduce a BC break by renaming it, so we'll live with this incoherence.

**Comment by fzaninotto** [comment]
Can you please remove the changes that aren't related to your new feature? The ones you cherry-picked related ti WizardForm for example.

**Comment by gitstart** [comment]
> Can you please remove the changes that aren't related to your new feature? The ones you cherry-picked related to WizardForm for example.

@fzaninotto Seems like other changes are from the `next` branch

**Comment by fzaninotto** [comment]
I think you messed up your rebase. 

Let me reiterate: you need to remove the changes you didn't make from this PR.

**Comment by fzaninotto** [comment]
Bump, can you please fix your rebase?

**Comment by fzaninotto** [comment]
No news for some time, closing.

**Pull request closed without merging**